### ISSUE TITLE
`(^.|x)y` rather slow

### ISSUE THREAD

**Issue by xrat** [Genivia/ugrep#426]
I was surprised to find that ugrep takes rather long to match `(^.|x)y`:

``` bash
$ ugrep --version | head -n 1
ugrep 6.5.0 x86_64-pc-linux-gnu +sse2; -P:pcre2jit
$ cat /etc/debian_version
11.11
$ ls -lh sample.txt
-rw------- 1 xrat users 42M 2024-09-24 22:49 sample.txt
$ file sample.txt
sample.txt: ASCII text
$ time ugrep -E -c '(^.|x)y' sample.txt
0

real	0m1.589s
user	0m1.589s
sys	0m0.000s

$ grep --version | head -n 1
grep (GNU grep) 3.6
@W:~$ time grep -E -c '(^.|x)y' sample.txt
0

real	0m0.084s
user	0m0.076s
sys	0m0.008s
```

AFAICT, it's not related to the input. I tried several. My `sample.txt` are 10 concatenated copies of `/usr/lib/x86_64-linux-gnu/perl/5.32.1/CORE/charclass_invlists.h`.

**Comment by genivia-inc** [comment]
The `^` anchor is handled differently than other anchors and word boundaries.

I'm not surprised it is slow in this special case, because letter `y` matches a lot and that makes it slower to match to filter out the cases when it should match only after one Unicode character at the start of a line.

Use option `-U` to speed this up. The `.` dot then only matches a single byte, not a Unicode wide char in UTF-8.

**Comment by xrat** [comment]
I tried `-U` and `LC_ALL=C` but there was not speedup.

I don't think it's directly related to `^.` because `^.` is fast, `(^.)` is still fast, but `(^.)x` is slow.

**Comment by genivia-inc** [comment]
Hm, it depends. Searching a large 100MB file the following way is fast:
```
$ ug -c '^.y' enwik8 --stats
1599
Searched 1 file in 0.0398 seconds: 1 matching (100%)
```
but this is slow:
```
$ ug -c '(^.)y' enwik8 --stats
1599
Searched 1 file in 2.54 seconds: 1 matching (100%)
```
It's caused by the `^` placement within parenthesis, not the `.y` matching per se, because that is fast for so many matches it returns:
```
$ ug -c '.y' enwik8 --stats
323401
Searched 1 file in 0.21 seconds: 1 matching (100%)
```
I'm surprised. Even though I wrote all the source code, I may have missed an opportunity to make sure `^` works fast in more cases.

The `-U` doesn't make a difference, because the `^` placement is unrelated to that.

**Comment by genivia-inc** [comment]
Quick update: working on an update, but further work has to wait until I'm back in a week. The update will include speed improvements that I made recently. This problem case triggers a worse-case scenario in which accelerated search is completely disabled (due to the anchor being up front) and also some backtracking takes place to try to match the anchor. This can and will be fixed.

**Comment by genivia-inc** [comment]
Work in progress report. Still need to work on some details.

For comparison, here are some interesting test cases after the updates that I worked on this week:
~~~
$ ugrep -c '^.y' enwik8 --stats
1599
Searched 1 file in 0.0353 seconds: 1 matching (100%)
~~~
This test case is much better than before after the update (see my earlier post for the old timing):
~~~
$ ugrep -c '(^.)y' enwik8 --stats
1599
Searched 1 file in 0.0407 seconds: 1 matching (100%)
~~~
Although PCRE2 (ugrep option `-P`) runs a bit faster for this test case:
~~~
$ ugrep -P -c '(^.)y' enwik8 --stats
1599
Searched 1 file in 0.0332 seconds: 1 matching (100%)
~~~
But GNU grep runs slower:
~~~
$ /usr/bin/time ggrep -c -E '(^.)y' enwik8
1599
        0.07 real         0.06 user         0.01 sys
~~~

Now, for the `(^.|x)y` pattern things are different. That's because we no longer require a match at the start of a line, only for `^.y` but not for `xy`. There are two approaches to implement this. One is to treat `^` as a `\n` character (invisible, so not part of the match), which will be very fast. The other is to check for a line boundary, which is obviously slower because for every match of `.y` we check if it is at a line boundary. There will be many cases when it is not at line boundary, so that makes this approach slow.

The ugrep implementation can match `\n`, but that is not done for `^`. So it is slower than it could be implemented:
~~~
$ ugrep -c '(^.|x)y' enwik8 --stats
3473
Searched 1 file in 0.884 seconds: 1 matching (100%)
~~~
If we rewrite this pattern by dropping the parenthesis, then the timing is better, but still not great:
~~~
$ ugrep -c '^.y|xy' enwik8 --stats
3473
Searched 1 file in 0.611 seconds: 1 matching (100%)
~~~
The timing is better, because the internal translation to a DFA is affected by the placement of anchors and word boundaries inside parenthesis, which in turn makes the search engine sub-optimal. Best is to avoid that, which is easy to do by expanding the regex by dropping the parenthesis.

PCRE2 is also sluggish with this one, but less so:
~~~
$ ugrep -P -c '(^.|x)y' enwik8 --stats
3473
Searched 1 file in 0.3 seconds: 1 matching (100%)
~~~
GNU grep is fast for this pattern, which suggests it internally searches `\n` for `^` perhaps?
~~~
$ /usr/bin/time ggrep -c -E '(^.|x)y' enwik8
3473
        0.07 real         0.06 user         0.01 sys
~~~
That optimization to search for `\n` makes sense, because doing the same with ugrep explicitly we get a fast response:
~~~
$ ugrep -c '(\n.|x)y' enwik8 --stats
3488
Searched 1 file in 0.0498 seconds: 1 matching (100%)
~~~
Note that the match count differs, because we match `\n` that results in different matching line counting logic going on internally.

Compared to the old results, the work-in-progress ugrep timings improved, so that is nice. But the specific cases of a regex `^` when combined with other non-`^` regex is not as fast as GNU grep when patterns are short. We know why that is so. The question is, is it really necessary to update the ugrep internals for `^` just for these cases of short one or two character long patterns? Note that longer patterns will match very fast and don't benefit from optimizing `^` with `\n` matching.
